diag_manager/fms_diag_axis_object.F90

```diff
@@ -275,6 +286,100 @@ subroutine set_axis_domain(obj, Domain, Domain2, DomainU)
     end select
   end subroutine set_axis_domain
 
+  subroutine fms_diag_axis_object_init()
+
+    if (module_is_initialized) return
+
+    number_of_axis = 0
+    allocate(axis_obj(max_axes))
+
+    module_is_initialized = .true.
+  end subroutine fms_diag_axis_object_init
+
+  subroutine fms_diag_axis_object_end()
+    deallocate(axis_obj)
+
+    module_is_initialized = .false.
+  end subroutine fms_diag_axis_object_end
+
+  !> @brief Wrapper for the register_diag_axis subroutine. This is needed to keep the diag_axis_init
+  !! interface the same
+  !> @return Axis id
+  FUNCTION modern_diag_axis_init(axis_name, axis_data, units, cart_name, long_name, direction,&
+    & set_name, edges, Domain, Domain2, DomainU, aux, req, tile_count, domain_position ) &
+    & result(id)
+
+    CHARACTER(len=*),   INTENT(in)           :: axis_name       !< Name of the axis
+    REAL,               INTENT(in)           :: axis_data(:)    !< Array of coordinate values
+    CHARACTER(len=*),   INTENT(in)           :: units           !< Units for the axis
+    CHARACTER(len=1),   INTENT(in)           :: cart_name       !< Cartesian axis ("X", "Y", "Z", "T", "U", "N")
+    CHARACTER(len=*),   INTENT(in), OPTIONAL :: long_name       !< Long name for the axis.
+    CHARACTER(len=*),   INTENT(in), OPTIONAL :: set_name        !< Name of the parent axis, if it is a subaxis
+    INTEGER,            INTENT(in), OPTIONAL :: direction       !< Indicates the direction of the axis
+    INTEGER,            INTENT(in), OPTIONAL :: edges           !< Axis ID for the previously defined "edges axis"
+    TYPE(domain1d),     INTENT(in), OPTIONAL :: Domain          !< 1D domain
+    TYPE(domain2d),     INTENT(in), OPTIONAL :: Domain2         !< 2D domain
+    TYPE(domainUG),     INTENT(in), OPTIONAL :: DomainU         !< Unstructured domain
+    CHARACTER(len=*),   INTENT(in), OPTIONAL :: aux             !< Auxiliary name, can only be <TT>geolon_t</TT>
+                                                                !! or <TT>geolat_t</TT>
+    CHARACTER(len=*),   INTENT(in), OPTIONAL :: req             !< Required field names.
+    INTEGER,            INTENT(in), OPTIONAL :: tile_count      !< Number of tiles
+    INTEGER,            INTENT(in), OPTIONAL :: domain_position !< Domain position, "NORTH" or "EAST"
+    integer :: id
+
+    number_of_axis = number_of_axis + 1
+
+    call axis_obj(number_of_axis)%register(axis_name, axis_data, units, cart_name, long_name=long_name, &
+    & direction=direction, set_name=set_name, edges=edges, Domain=Domain, Domain2=Domain2, DomainU=DomainU, aux=aux, &
+    & req=req, tile_count=tile_count, domain_position=domain_position)
+
+    id = number_of_axis
+  end function
+
+  !> @brief Check if a cart_name is valid and crashes if it isn't
+  subroutine check_if_valid_cart_name(cart_name)
+    character(len=*), intent(in) :: cart_name
+
+    select case (cart_name)
+    case ("X", "Y", "Z", "T", "U", "N")
```

Comment:
These two codes do the exact same thing. The way it is now is cleaner ...

Reply:
The two codes don't do exactly the same thing. 
 
Your code creates 2 branches 
First it will evaluate if cart_name is X, Y, Z, T, U, N.  If it is any of those values, it will do nothing.  So you've done that evaluation for no reason... an empty branch.  If it doesn't match any of those values, it will do essentially an else and then hit your error. 

The `if` statement code does one evaluation with no else.  Even though it's "messier", it's more efficient and only creates a single branch.  If you can represent a single branch in a `case` statement with this code, then go for it.

Reply:
i think, you code should be AND not OR (cartname="Y" will crash because it is not "X")

`First it will evaluate if cart_name is X, Y, Z, T, U, N. If it is any of those values, it will do nothing.`

Yeah thats the point as you soon as my code determines that `cartname` is either X, Y, Z, T, U, N it will do nothing and leave the subroutine. Your code is going to check that `cartname` is not X AND  not Y AND not Z AND not T AND not U AND not N  if all that is true then it crash otherwise it will do nothing and leave the subroutine. So technically my code is more "efficient"...